src/colmap/estimators/covariance.h

```diff
@@ -0,0 +1,178 @@
+
+  // Interfaces
+  // test if the block corresponds to any parameter in the problem except for 3D
+  // points
+  bool HasBlock(const double* params) const;
+  // test if the pose is inside the problem as non-constant variables
+  bool HasPose(image_t image_id) const;
+
+  // pose parameters
+  Eigen::MatrixXd GetPoseCovariance() const;
+  Eigen::MatrixXd GetPoseCovariance(image_t image_id) const;
+  Eigen::MatrixXd GetPoseCovariance(
+      const std::vector<image_t>& image_ids) const;
+  Eigen::MatrixXd GetPoseCovariance(image_t image_id1, image_t image_id2) const;
+
+  // all parameters (except for 3D points)
+  Eigen::MatrixXd GetCovariance(double* parameter_block) const;
+  Eigen::MatrixXd GetCovariance(
+      const std::vector<double*>& parameter_blocks) const;
+  Eigen::MatrixXd GetCovariance(double* parameter_block1,
+                                double* paramter_block2) const;
+
+  // test if either ``ComputeFull()`` or ``Compute()`` has been called
+  bool HasValidPoseCovariance() const;
+  // test if ``ComputeFull()`` has been called
+  bool HasValidFullCovariance() const;
+
+ protected:
+  // blocks parsed from reconstruction (initialized at construction)
+  std::vector<const double*> pose_blocks_;
+  int num_params_poses_ = 0;
+  std::vector<const double*> other_variables_blocks_;
+  int num_params_other_variables_ = 0;
+  std::vector<const double*> point_blocks_;
+  int num_params_points_ = 0;
+
+  // get the starting index of the parameter block in the matrix
+  // orders: [pose_blocks, other_variables_blocks, point_blocks]
+  std::map<const double*, int> map_block_to_index_;
+  int GetBlockIndex(const double* params) const;
+  int GetBlockTangentSize(const double* params) const;
+  int GetPoseIndex(image_t image_id) const;
+  int GetPoseTangentSize(image_t image_id) const;
+
+  // covariance for all parameters (except for 3D points), equal to
+  // H_cc_.inverse()
+  Eigen::MatrixXd cov_variables_;
+
+  // covariance for pose parameters
+  Eigen::MatrixXd cov_poses_;
+
+  // ceres problem
+  ceres::Problem* problem_;
+
+  // reconstruction
+  Reconstruction* reconstruction_;
+
+ private:
+  // set up parameter blocks
+  void Setup();
+};
+
+class BundleAdjustmentCovarianceEstimatorCeresBackend
+    : public BundleAdjustmentCovarianceEstimatorBase {
+ public:
+  BundleAdjustmentCovarianceEstimatorCeresBackend(
+      ceres::Problem* problem, Reconstruction* reconstruction)
+      : BundleAdjustmentCovarianceEstimatorBase(problem, reconstruction) {}
+
+  bool ComputeFull() override;
+  bool Compute() override;
+};
+
+class BundleAdjustmentCovarianceEstimator
+    : public BundleAdjustmentCovarianceEstimatorBase {
+ public:
+  BundleAdjustmentCovarianceEstimator(ceres::Problem* problem,
+                                      Reconstruction* reconstruction)
+      : BundleAdjustmentCovarianceEstimatorBase(problem, reconstruction) {}
+
+  void SetLambda(double lambda);  // update daming factor. default = 1e-6
```

Comment:
Instead of a separate method, we could also just expect it in the constructor and keep it const?

Reply:
Sounds good. This has been done.